### ISSUE TITLE
Segmentation fault in ckch_deinit () at src/ssl_ckch.c:3858 for HAProxy 2.6.x

### ISSUE THREAD

**Issue by atoom** [haproxy/haproxy#1836]
### Detailed Description of the Problem

Hi,

We are upgrading our HAProxy installation from 2.4.4 to 2.6.3 and getting segmentation fault when checking our configuration via `haproxy -f haproxy.cfg -c` using the same configuration input that previously worked on version 2.4.4.

We are running the alpine docker images with a bind-mount of the host filesystem for all the config.

There seems to be a lot of randomness in this issue. Simply deleting/renaming files/directories causes the segfault to occur/dissapear. Unfortunately it is related to ssl so I cannot submit the full input data since it involves key material etc.

```
# haproxy -f /opt/haproxy-prod/conf/haproxy.cfg.minimal -c
[NOTICE]   (249) : haproxy version is 2.6.3-76f187b
[WARNING]  (249) : config : ca-file: 0 CA were loaded from '/opt/haproxy-prod/conf/ca/xxx/'
[WARNING]  (249) : config : ca-file: 0 CA were loaded from '/opt/haproxy-prod/conf/ca/yyy/'
[WARNING]  (249) : config : ca-file: 0 CA were loaded from '/opt/haproxy-prod/conf/ca/zzz/'
Warnings were found.
Configuration file is valid
Segmentation fault (core dumped)
```

### Expected Behavior

haproxy should exit without any issues

### Steps to Reproduce the Behavior

1. docker run --name haproxy-core-dump --privileged -it -v /tmp/haproxy/haproxy-core-dump/opt/haproxy-prod:/opt/haproxy-prod haproxytech/haproxy-alpine:2.6.3 sh
2. haproxy -f /opt/haproxy-prod/conf/haproxy.cfg.minimal -c

### What is your configuration?

```haproxy
global
    log           stdout format raw local0

    chroot        /var/lib/haproxy
    pidfile       /var/run/haproxy.pid
    stats socket  /var/run/haproxy.stats
    maxconn       10000
    user          haproxy
    group         haproxy
    master-worker

defaults
    mode                    http
    log                     global
    timeout http-request    5s
    timeout queue           1m
    timeout connect         10s
    timeout client          1m
    timeout server          1m
    timeout http-keep-alive 10s
    timeout check           10s
    maxconn                 10000

frontend fe-http-1
    bind :48292 ssl crt /opt/haproxy-prod/conf/ssl ca-file /opt/haproxy-prod/conf/ca/xxx/ verify required

frontend fe-http-2
    bind :48291 ssl crt /opt/haproxy-prod/conf/ssl ca-file /opt/haproxy-prod/conf/ca/yyy/ verify required

frontend fe-http-3
    bind :48293 ssl crt /opt/haproxy-prod/conf/ssl ca-file /opt/haproxy-prod/conf/ca/zzz/ verify required
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.6.3-76f187b 2022/08/19 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.3.html
Running on: Linux 5.4.72-microsoft-standard-WSL2 #1 SMP Wed Oct 28 23:40:43 UTC 2020 x86_64
Build options :
  TARGET  = linux-musl
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_LINUX_TPROXY=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_SLZ=1 USE_TFO=1 USE_PROMEX=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL +THREAD -BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM -ZLIB +SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.1q  5 Jul 2022
Running on OpenSSL version : OpenSSL 1.1.1q  5 Jul 2022
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.40 2022-04-14
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 10.3.1 20211027

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
        [CACHE] cache
        [COMP] compression
        [FCGI] fcgi-app
        [SPOE] spoe
        [TRACE] trace
```

### Last Outputs and Backtraces

```plain
Reading symbols from haproxy...

warning: Can't open file /dev/zero (deleted) during file-backed mapping note processing
[New LWP 19]
Core was generated by `haproxy -f haproxy.cfg.minimal -c'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  eb_next (node=0x7f4cd5ab6738) at include/import/ebtree.h:378
378     include/import/ebtree.h: No such file or directory.
(gdb) t a a bt full

Thread 1 (LWP 19):
#0  eb_next (node=0x7f4cd5ab6738) at include/import/ebtree.h:378
        t = 0x7f4cd5ad05d9
        t = <optimized out>
#1  ebmb_next (ebmb=0x7f4cd5ab6738) at include/import/ebmbtree.h:51
No locals.
#2  ckch_deinit () at src/ssl_ckch.c:3858
        entry = 0x7f4cd5ab6710
        node = <optimized out>
        next = <optimized out>
        store = <optimized out>
        canode = 0x7f4cd5ab6738
#3  0x00005611901deb95 in deinit () at src/haproxy.c:2672
        p = <optimized out>
        p0 = <optimized out>
        wl = <optimized out>
        wlb = <optimized out>
        uap = <optimized out>
        ua = <optimized out>
        log = <optimized out>
        logb = <optimized out>
        bol = <optimized out>
        bolb = <optimized out>
        pdf = 0x7f4cd5cf57b0
        pdfb = <optimized out>
        pxdf = <optimized out>
        pxdfb = <optimized out>
        srvdf = <optimized out>
        srvdfb = <optimized out>
        tif = <optimized out>
        tifb = <optimized out>
        tdf = <optimized out>
        tdfb = <optimized out>
        taf = <optimized out>
        tafb = <optimized out>
        tff = <optimized out>
        tffb = <optimized out>
        pscf = <optimized out>
        pscfb = <optimized out>
        pcf = <optimized out>
        pcfb = <optimized out>
        ppcf = <optimized out>
        ppcfb = <optimized out>
        prcf = <optimized out>
        prcfb = <optimized out>
        pprs = <optimized out>
        pprsb = <optimized out>
        cur_fd = <optimized out>
#4  0x00005611901df202 in deinit_and_exit (status=status@entry=0) at src/haproxy.c:2775
No locals.
#5  0x000056119009b8ba in init (argc=<optimized out>, argv=<optimized out>) at src/haproxy.c:2215
        pr = <optimized out>
        px = <optimized out>
        px = <optimized out>
        pcf = <optimized out>
        progname = <optimized out>
        err_code = <optimized out>
        wl = <optimized out>
        prcf = <optimized out>
        ideal_maxconn = <optimized out>
        progname = <optimized out>
        err_code = <optimized out>
        wl = <optimized out>
        px = <optimized out>
        pcf = <optimized out>
        prcf = <optimized out>
        ideal_maxconn = <optimized out>
        result = <optimized out>
        err = <optimized out>
        errptr = <optimized out>
        errmsg = <optimized out>
        args = <optimized out>
        arg = <optimized out>
        outlen = <optimized out>
        w = <optimized out>
        env_cfgfiles = <optimized out>
        env_err = <optimized out>
        ret = <optimized out>
        tmproc = <optimized out>
        it = <optimized out>
        c = <optimized out>
        __ret = <optimized out>
        srv = <optimized out>
        ppcf = <optimized out>
        pscf = <optimized out>
        pr = <optimized out>
        px = <optimized out>
        p = <optimized out>
        sides = <optimized out>
        mem = <optimized out>
        retried = <optimized out>
        sides = <optimized out>
        mem = <optimized out>
        sslmem = <optimized out>
        sides = <optimized out>
        mem = <optimized out>
        clearmem = <optimized out>
        retried = <optimized out>
        s = <optimized out>
#6  main (argc=<optimized out>, argv=0x7ffe7289c2a8) at src/haproxy.c:3084
        err = <optimized out>
        retry = <optimized out>
        limit = {rlim_cur = 1048576, rlim_max = 1048576}
        pidfd = -1
        intovf = <optimized out>
        msg = <optimized out>
```

**Comment by capflam** [comment]
Sorry, I referenced the wrong issue in the commit. It should be #1839 ....

**Comment by wlallemand** [comment]
"using the same configuration input that previously worked on version 2.4.4." Well, I don't see how it's possible since the support for directories on the ca-file directive was added in 2.6.

Your issue seems related to the absence of files in the directories, I could reproduce when built with ASAN. In fact it should trigger an error in this case, not a warning, such a configuration is not valid without files!

**Comment by atoom** [comment]
The example I provided is a simplified configuration where I have tried to remove as many input sources as possible to try to narrow down the search scope of the error.

I just copied our 2.4.4 production configuration + certs to my local machine and ran the config check using haproxy 2.6.4 and I still get the same error.

Like I mentioned in the initial description the error appears/disappears in a minimal configuration if I shift around the paths or add/remove some certs in the different directories. I have a local src version of haproxy running locally so I am able to add some debug print statements around the code but I am not really sure what to look for. Right now I am just trying to print different memory adresses to see if there is a double free. I would really appreciate suggestions on what to debug log so I can provide more details to this case!

**Comment by wlallemand** [comment]
Okay, I was confused because of the warnings in your output.
It looks like I could also trigger the problem with a valid configuration and ASAN. It still a problem with the ca-file though, but I have enough information with ASAN to fix the issue:

```$ ./haproxy -f issue-1836.cfg -c
Configuration file is valid
=================================================================
==95904==ERROR: AddressSanitizer: heap-use-after-free on address 0x60800000e458 at pc 0x00000057243f bp 0x7ffdff5ec390 sp 0x7ffdff5ec388
READ of size 8 at 0x60800000e458 thread T0
    #0 0x57243e in eb_next /home/wla/projects/haproxy/haproxy-community-maint/include/import/ebtree.h:506:48
    #1 0x5722c5 in ebmb_next /home/wla/projects/haproxy/haproxy-community-maint/include/import/ebmbtree.h:51:9
    #2 0x5722c5 in ckch_deinit /home/wla/projects/haproxy/haproxy-community-maint/src/ssl_ckch.c:3910:12
    #3 0x529d61 in ssl_sock_deinit /home/wla/projects/haproxy/haproxy-community-maint/src/ssl_sock.c:5595:2
    #4 0xbcc1b2 in deinit /home/wla/projects/haproxy/haproxy-community-maint/src/haproxy.c:2683:3
    #5 0xbcefe2 in deinit_and_exit /home/wla/projects/haproxy/haproxy-community-maint/src/haproxy.c:2786:2
    #6 0xbd910a in init /home/wla/projects/haproxy/haproxy-community-maint/src/haproxy.c:2223:4
    #7 0xbd1687 in main /home/wla/projects/haproxy/haproxy-community-maint/src/haproxy.c:3112:2
    #8 0x7f9844938d8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #9 0x7f9844938e3f in __libc_start_main csu/../csu/libc-start.c:392:3
    #10 0x447ef4 in _start (/home/wla/projects/haproxy/haproxy-community-maint/haproxy+0x447ef4)

0x60800000e458 is located 56 bytes inside of 91-byte region [0x60800000e420,0x60800000e47b)
freed by thread T0 here:
    #0 0x4c4b72 in __interceptor_free (/home/wla/projects/haproxy/haproxy-community-maint/haproxy+0x4c4b72)
    #1 0x56e5c8 in ssl_store_delete_cafile_entry /home/wla/projects/haproxy/haproxy-community-maint/src/ssl_ckch.c:1164:2
    #2 0x5722d2 in ckch_deinit /home/wla/projects/haproxy/haproxy-community-maint/src/ssl_ckch.c:3911:3
    #3 0x529d61 in ssl_sock_deinit /home/wla/projects/haproxy/haproxy-community-maint/src/ssl_sock.c:5595:2
    #4 0xbcc1b2 in deinit /home/wla/projects/haproxy/haproxy-community-maint/src/haproxy.c:2683:3
    #5 0xbcefe2 in deinit_and_exit /home/wla/projects/haproxy/haproxy-community-maint/src/haproxy.c:2786:2
    #6 0xbd910a in init /home/wla/projects/haproxy/haproxy-community-maint/src/haproxy.c:2223:4
    #7 0xbd1687 in main /home/wla/projects/haproxy/haproxy-community-maint/src/haproxy.c:3112:2
    #8 0x7f9844938d8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16

previously allocated by thread T0 here:
    #0 0x4c4f72 in __interceptor_calloc (/home/wla/projects/haproxy/haproxy-community-maint/haproxy+0x4c4f72)
    #1 0x56d9a4 in ssl_store_create_cafile_entry /home/wla/projects/haproxy/haproxy-community-maint/src/ssl_ckch.c:1069:9
    #2 0x56f45d in ssl_store_load_locations_file /home/wla/projects/haproxy/haproxy-community-maint/src/ssl_ckch.c:1363:10
    #3 0xc71a95 in httpclient_precheck /home/wla/projects/haproxy/haproxy-community-maint/src/http_client.c:1193:8
    #4 0xbd8564 in init /home/wla/projects/haproxy/haproxy-community-maint/src/haproxy.c:2123:15
    #5 0xbd1687 in main /home/wla/projects/haproxy/haproxy-community-maint/src/haproxy.c:3112:2
    #6 0x7f9844938d8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16

SUMMARY: AddressSanitizer: heap-use-after-free /home/wla/projects/haproxy/haproxy-community-maint/include/import/ebtree.h:506:48 in eb_next
Shadow bytes around the buggy address:
  0x0c107fff9c30: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c107fff9c40: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c107fff9c50: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9c60: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c107fff9c70: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
=>0x0c107fff9c80: fa fa fa fa fd fd fd fd fd fd fd[fd]fd fd fd fd
  0x0c107fff9c90: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c107fff9ca0: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9cb0: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c107fff9cc0: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c107fff9cd0: fa fa fa fa fd fd fd fd fd fd fd fd fd fd fd fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==95904==ABORTING
```

**Comment by capflam** [comment]
Backported to 2.6.

**Comment by atoom** [comment]
Thanks for the fix - I have tested the 2.6 branch against our production configuration and there is no more segfault!
